### ISSUE TITLE
Just a heads up on a new spoofing add-on

### ISSUE THREAD

**Issue by laserstrike** [ghacksuserjs/ghacks-user.js#203]
https://github.com/snapper26/shapeshifter
https://addons.mozilla.org/en-US/firefox/addon/shape-shifter/

According to the .js user agents shouldn't be spoofed, but I found this add-on that does a bit more than that and I just wanted to know your thoughts on it. Thanks.

**Comment by Thorin-Oakenpants** [comment]
Thanks @lazerstrike

`Generates randomised values` ... `No configuration needed`

Several issues with this (haven't had time to look at anything)
- No control over the spoofing list which can lead to site breakage. That said, if it limits itself to 
Firefox, that may be OK.
- resistFP will probably override this (FF56)
- does this cover buildID, because even though the spoof is random, if the buildID doesn't match a one of those for that version & OS, then it sticks out
- I am firmly in the camp that it is far better, as per TBB's draft, where possible, to lower entropy vs randomly-always raising it
- does this pass the test for all the navigator leaks and over iframes etc
- more to come, watch this space --> <-- this one

**Comment by crssi** [comment]
The real browser is revealed over iFrame method. So no use of shapeshifter... yet.

**Comment by Snapper26** [comment]
Hi there
I am the author of this addon. Thank you for your feedback on this matter. I had a sneaking suspicion that <iframe> would cause a leakage however it is impossible to catch all problems at once. I have just released this addon and have plans to fix the issues mentioned above with <iframe> as well as another issue with CSP (Content Security Policy) which causes one of my injected scripts to be blocked from being injected causing the spoofing to fail. I try to normalise things like the navigator to common values in order to blend in.

As for buildID, I have only spoofed properties that appear under the standard list here https://developer.mozilla.org/en-US/docs/Web/API/Navigator so far but am open to spoofing anyother properties that need to be spoofed. As for spoofing other things such as batterymanager, canvas, webgl, audiocontext and client rectangles, the reason is because I believe them to be fingerprinting vectors that if left untouched could render the addon pointless. It is hard to give something like a canvas image a "default" value (maybe transparent black?) so what I do is generate a seed for each session and each different domain/origin which is used to create a unique ()but consistent) canvas for that session, it then changes when the session changes.

**Comment by Thorin-Oakenpants** [comment]
Hey @Snapper26 : ) Welcome

What I wrote above was a quick 2 minutes of thought off the top of my head, followed by some edited points after I had a quick look at the repo readme. Nothing is meant to be be dismissive, just ideas for discussion / follow-up. Rome wasn't built in a day :) 

Some random thoughts below...

---

> As for buildID, I have only spoofed properties that appear under the standard list...

Right. So the buildID leaks which gives away your exact version and OS. I've have to check now, but I think Tor Uplift just blanks it (there's actually a FF pref that does this), but that means you are now a KNOWN spoofer. Tor Uplift doesn't care because they just want uniformity (enforced). If 1% of FF's 100 million users turn on resistFingerprinting (there will be a UI option for it at some stage, and I expect it to even be added to PB mode windows down the track), then that is 1 million users. That's a good start.

Sticking to UA only, I think there are too many holes and overlap. That said, those holes **may** or **may not** exist (eg tor uplift navigator spoof may close some of these but the vast bulk of browser IDing remains - eg feature detect, eg CSS line height can detect OS, scrollbars can do the same) whether you spoof or not. If you're always rapidly changing then who cares....  there's no right or wrong answer - except the value of being a spoofer or not is itself a value - i.e if you are caught out in a lie then you are unique, otherwise they do not know you spoofed. I call this creating a problem that didn't exist, or making it more complex :)

Tor Uplift in FF56+ enforces spoofed values for UA and for items relating to UA. Have you checked if your extension overrides theirs - I will assume it does, as script injection should come last, right?. If it doesn't then that's a major problem.

---

I do not think referers should be in this AT ALL, and people can already use extensions such as uMatrix spoof all, or dedicated web ext such as smart referer etc - bit of a transition going on between legacy and WebExt, so its a bit messy. Re-inventing the wheel and making your addon more complex.

For this reason alone - interfering with referers and giving me no control - I would not use your extension: i.e granular control is the best solution, and this extension screws that up, so to speak. At the very least options down the track to turn features on and off.

---

Long story short .. I am not an expert, anti-FPing is really really hard, there are often no right or wrong answers... the Web Ext is at initial release and will prototype as we go ... good to have you around :)

**Comment by Snapper26** [comment]
As of 12 August 2017 shapeshifter uses the value "Mozilla/5.0 (Windows NT 6.1; rv:52.0) Gecko/20100101 Firefox/52.0" for navigator.userAgent which means that the build ID should be "20100101"? If so then I have fixed this today and made navigator.buildID equal to "20100101". If the user agent changes then this might need to change too. As for the referrer, sometimes it is good to get usability feedback which can help guide you in the right direction. I would like to in the not too distant future add a proper HTML pop up GUI which more detailed configuration options. At the moment I am just going to focus on fixing the issues brought up above such as buildID, iframes and the CSP script injection issue I am facing.

**Comment by Thorin-Oakenpants** [comment]
navigator.buildID -> http://browserspy.dk/showprop.php -> mine is 20170803103124 (I am not hiding it) => windows FF55 time stamp (seems weird its naming convention indicates august 3rd)

`general.buildID.override` overrides this in FF and the consensus has always been to use `20100101` (because the pref landed in 2010, you see). This is a simple test to see if someone is faking their agent :)

A more comprehensive approach (and I'm not saying to do it) is to incorporate the actual real values into every UA used. This adds complexity and overhead and time wasting ... not that I don't applaud such efforts :)  If you spoof as Linux version 54 language UK - then the build id should match linux 54 (and language could be English-UK, timestamp could be UK, date formats should be uk not that backwards-assed american format, and so on). That said, we already know you;'re spoofing, so big deal - its randomized frequently - **the plausibility matters way less than the values, in fact plausibility might even be zero - I wouldn't get too hung on this aspect** :)

If I think of things I will bring them up in your repo - I'm already a watcher

**Comment by Snapper26** [comment]
So if 20100101 gives me away as trying to hide, then that is a fingerprint in itself. I know you didn't say to do it, however it might be worth the extra bit of effort to use the real buildID for each release. This would make it much more believable. Also I have already spend lots of time making this extension so yeah.

> If I think of things I will bring them up in your repo - I'm already a watcher

This would be greatly appreciated as sometimes I get so deep in development that I miss the simple things.

**Comment by Atavic** [comment]
Leave it so, most UA Strings contain it. A partial list is [here](www.useragentstring.com/pages/useragentstring.php?name=Firefox) and a huge one [here](https://myusers.info/client/firefox?p=370).

Also recent Pale-Moon and K-Meleon give `20100101`

**Comment by Atavic** [comment]
Leave it so, most UA Strings contain it. A partial list is [here](http://www.useragentstring.com/pages/useragentstring.php?name=Firefox) and a huge one [here](https://myusers.info/client/firefox?p=370)
.

Also recent Pale-Moon and K-Meleon give 20100101

**Comment by Atavic** [comment]
Most UA Strings contain it. A partial list is [here](http://www.useragentstring.com/pages/useragentstring.php?name=Firefox) and a huge one [here](https://myusers.info/client/firefox?p=370).

Also recent Pale-Moon and K-Meleon give `20100101`

**Comment by Snapper26** [comment]
Yep so I probably should just use 20100101 then? Also @crssi you mentioned about iframe.contentWindow still revealing the real browser attributes. Shape Shifter is a web extension which means that I can't use the obsolete/deprecated preferences service to set say general.useragent.override. This makes spoofing in iframe's quite tricky unless someone knows an easy way that I don't :)

**Comment by Thorin-Oakenpants** [comment]
Closing this. Am adding the extension to out addon sticky, so we do not forget about it.
